[DOCS] Adds docs to semantic text (#108311)
elastic/elasticsearch: docs/reference/query-dsl/semantic-query.asciidoc

Reviewer: This document also uses the term "inference service". Can we standardize on "inference endpoint"?

Reviewer 2: ++ - @szabosteve , can you please make that change and ensure it's consistent?

semantic search using `semantic_text` and `semantic` query.

[discrete]
[[hybrid-search-semantic]]
==== Hybrid search with the `semantic` query

The `semantic` query can be used as a part of a hybrid search where `semantic`
query is combined with lexical queries. For example, the query below finds
documents with title matching "mountain lake", and combines them with doing the
same semantic search on the field `title_semantic`, that is a `semantic_text`
field. The combined documents are then scored, and the top 3 top scored
documents are returned.

[source,console]
------------------------------------------------------------
POST my-index/_search
{
  "size" : 3,
  "query": {
    "bool": {
      "should": [
        {
          "match": {
            "title": {
              "query": "mountain lake",
              "boost": 1
            }
          }
        },
        {
          "semantic": {
            "field": "title_semantic",
            "query": "mountain lake",
            "boost": 2
          }
        }
      ]
    }
  }
}
------------------------------------------------------------
// TEST[skip:TBD]

[discrete]
[[search-sparse-inference]]

> ==== Search with `sparse_embedding` inference

Reviewer 2: Do you think it would be a good idea to add a paragraph explaining about the limitations of semantic query, that will serve as the intro for using sparse and dense inference searches?

Reviewer: ++

Author: @carlosdelest  Yes, we can add a separate section called `Limitations` to this page. Can you point me to a source of info so I know the content of such a section?

Reviewer: My opinion: We should add a section called "Customizing the `semantic` query" just before "Search with `sparse_embedding` inference". We can use this section to explain why we didn't add a bunch of params to the `semantic` query:

```
==== Customizing the `semantic` query

You may have noticed that the `semantic` query doesn't provide many top-level parameters for customization.
This is intentional; it is intended to be used as a stepping stone to simplify the setup & evaluation of semantic search.
Once you have a good understanding of how to best apply semantic search in your use case, we recommend graduating to using the `sparse_vector` or `knn` query, depending on the inference type.
These query types provide greater flexibility, allowing you to customize behavior for better relevance and/or performance.
```

@carlosdelest WDYT?

Reviewer 2: We had an intro text at the [original doc](https://docs.google.com/document/d/1DYhE6tjw6Bdh1Ctr2-5S-FBW7ZhV5El-jia78i07bMc/edit#heading=h.en3pyq5aexc6) we used. I'm happy with a mix of that and Mike's paragraph in the example.

I trust your judgement here @szabosteve !

Reviewer: I like the text in the original doc as well, and I think "Advanced search on semantic_text fields" is a better heading. Combine as you see fit @szabosteve!

Author: Sorry, my mistake. I left our `Advanced search on semantic_text fields` by accident. I've added the section to my latest commit, combined the text with Mike's suggestion, and made it as concise as possible.

When the {infer} service uses a `sparse_embedding` model, you can use a

> `sparse_vector` query on a `semantic_text` field in the following way:

Reviewer 2: we need to add a link once it's merged

[source,console]
------------------------------------------------------------
GET test-index/_search
{
  "query": {
    "nested": {
      "path": "inference_field.inference.chunks",
      "query": {
        "sparse_vector": {
          "field": "inference_field.inference.chunks.embeddings",
          "inference_id": "my-inference-id",

>           "query_text": "mountain lake"

Reviewer 2: We changed the sparse_vector query text field

```suggestion
          "query": "mountain lake"
```

        }
      }
    }
  }
}
------------------------------------------------------------
// TEST[skip:TBD]

You can customize the `sparse_vector` query to include specific settings, like
pruning configuration.

[discrete]
[[search-text-inferece]]
==== Search with text_embedding inference

> When the {infer} service uses a `text_embedding` model, you can use a `knn`
> query on a `semantic_text` field in the following way:

Reviewer 2: ```suggestion
When the {infer} service uses a `text_embedding` model, you can use a <<query-dsl-knn-query,`knn`query>> on a `semantic_text` field in the following way:
```

[source,console]
------------------------------------------------------------
GET test-index/_search

> {
>   "query": {
>     "knn": {
>       "field": "inference_field",
>       "query_vector_builder": {
>         "text_embedding": { 
>           "model_id": "my_inference_id", 
>           "model_text": "mountain lake" 
>         }
>       }
>     }
>   }
> }

Reviewer 2: ```suggestion
{
  "query": {
    "nested": {
      "path": "inference_field.inference.chunks",
      "query": {
        "knn": {
          "field": "inference_field.inference.chunks.embeddings",
          "query_vector_builder": {
            "text_embedding": {
              "model_id": "my_inference_id",
              "model_text": "mountain lake"
            }
          }
        }
      }
    }
  }
}
```

[[query-dsl-semantic-query]]
=== Semantic query
++++
<titleabbrev>Semantic</titleabbrev>
++++

beta[]

The `semantic` query type enables you to perform semantic search on data stored in a <<semantic-text,`semantic_text`>> field.

[discrete]
[[semantic-query-example]]
==== Example request

[source,console]
------------------------------------------------------------
GET my-index-000001/_search
{
  "query": {
    "semantic": {
      "field": "inference_field",
      "query": "Best surfing places"
    }
  }
}
------------------------------------------------------------
// TEST[skip:TBD]

[discrete]
[[semantic-query-params]]
==== Top-level parameters for `semantic`

field::
(Required, string)
The `semantic_text` field to perform the query on.

query::
(Required, string)
The query text to be searched for on the field.

Refer to <<semantic-search-semantic-text,this tutorial>> to learn more about semantic search using `semantic_text` and `semantic` query.

[discrete]
[[hybrid-search-semantic]]
==== Hybrid search with the `semantic` query

> The `semantic` query can be used as a part of a hybrid search where `semantic` query is combined with lexical queries.

Reviewer: Missing "the"?: "where the `semantic` query"

> For example, the query below finds documents with title matching "mountain lake", and combines them with doing the same semantic search on the field `title_semantic`, that is a `semantic_text` field.

Reviewer: Suggestion: "and combines them with results from a semantic search on the field"

> For example, the query below finds documents with title matching "mountain lake", and combines them with doing the same semantic search on the field `title_semantic`, that is a `semantic_text` field.

Reviewer 2: ```suggestion
For example, the query below finds documents with title matching "mountain lake", and combines them with results from a semantic search on the field `title_semantic`, that is a `semantic_text` field.
```

[source,console]
------------------------------------------------------------
POST my-index/_search
{
  "size" : 3,
  "query": {
    "bool": {
      "should": [
        {
          "match": {
            "title": {
              "query": "mountain lake",
              "boost": 1
            }
          }
        },
        {
          "semantic": {
            "field": "title_semantic",
            "query": "mountain lake",
            "boost": 2
          }
        }
      ]
    }
  }
}
------------------------------------------------------------
// TEST[skip:TBD]

You can also use semantic_text as part of <<rrf,Reciprocal Rank Fusion>>:

[source,console]
------------------------------------------------------------
GET my-index/_search
{
  "retriever": {
    "rrf": {
      "retrievers": [
        {
          "standard": {
            "query": {
              "term": {
                "text": "shoes"
	            }
	          }
	        }
	      },
	      {
          "semantic": {
            "field": "semantic_field",
	          "query": "shoes"
	        }
	      }
	    ],
	    "rank_window_size": 50,
	    "rank_constant": 20
	  }
	}

> }

Reviewer: Nit: the indentation is off in this example, makes it harder to read

Reviewer 2: Hopefully fixed below

> {
>   "retriever": {
>     "rrf": {
>       "retrievers": [
>         {
>           "standard": {
>             "query": {
>               "term": {
>                 "text": "shoes"
> 	            }
> 	          }
> 	        }
> 	      },
> 	      {
>           "semantic": {
>             "field": "semantic_field",
> 	          "query": "shoes"
> 	        }
> 	      }
> 	    ],
> 	    "rank_window_size": 50,
> 	    "rank_constant": 20
> 	  }
> 	}
> }

Reviewer 2: ```suggestion
{
  "retriever": {
    "rrf": {
      "retrievers": [
        {
          "standard": {
            "query": {
              "term": {
                "text": "shoes"
              }
            }
          }
        },
        {
          "semantic": {
            "field": "semantic_field",
            "query": "shoes"
          }
        }
      ],
      "rank_window_size": 50,
      "rank_constant": 20
    }
  }
}
```

------------------------------------------------------------
// TEST[skip:TBD]

[discrete]
[[search-sparse-inference]]
==== Search with `sparse_embedding` inference

> When the {infer} endpoint uses a `sparse_embedding` model, you can use a `sparse_vector` query on a <<semantic-text,`semantic_text`>> field in the following way:

Reviewer 3: Link to sparse vector query docs here

Author: Added via [6a086bc](https://github.com/elastic/elasticsearch/pull/108311/commits/6a086bc02815060ae56a3320aea9e0ab23d413c9).

[source,console]
------------------------------------------------------------
GET test-index/_search
{
  "query": {
    "nested": {
      "path": "inference_field.inference.chunks",
      "query": {
        "sparse_vector": {
          "field": "inference_field.inference.chunks.embeddings",
          "inference_id": "my-inference-id",
          "query": "mountain lake"
        }
      }
    }
  }

> }

Reviewer: @carlosdelest will this query work as written given that `inference_field.inference.chunks` is a nested field?

Reviewer 2: I tested it. We should definitely find a way of testing the queries in this doc, hopefully David Kyle provides some insight on allowing this

------------------------------------------------------------
// TEST[skip:TBD]

> You can customize the `sparse_vector` query to include specific settings, like pruning configuration.

Reviewer 3: Maybe link to token pruning section of sparse vector docs here?

Author: I linked to the parameters section as the pruning section doesn't have an anchor.

> You can customize the `sparse_vector` query to include <<sparse-vector-field-params,specific settings>>, like pruning configuration.

Author: ```suggestion
You can customize the `sparse_vector` query to include specific settings, like <<sparse-vector-query-with-pruning-config-and-rescore-example,pruning configuration>>.
```

[[query-dsl-semantic-query]]
=== Semantic query
++++
<titleabbrev>Semantic</titleabbrev>
++++

beta[]

> The `semantic` query type enables you to perform semantic search on data stored in a <<semantic-text,`semantic_text`>> field.

Reviewer 2: ```suggestion
The `semantic` query type enables you to perform <<semantic-search,semantic search>> on data stored in a <<semantic-text,`semantic_text`>> field.
```

[discrete]
[[semantic-query-example]]
==== Example request

[source,console]
------------------------------------------------------------
GET my-index-000001/_search
{
  "query": {
    "semantic": {
      "field": "inference_field",
      "query": "Best surfing places"
    }
  }
}
------------------------------------------------------------
// TEST[skip:TBD]

[discrete]
[[semantic-query-params]]
==== Top-level parameters for `semantic`

field::
(Required, string)
The `semantic_text` field to perform the query on.

query::
(Required, string)
The query text to be searched for on the field.

Refer to <<semantic-search-semantic-text,this tutorial>> to learn more about semantic search using `semantic_text` and `semantic` query.

[discrete]
[[hybrid-search-semantic]]
==== Hybrid search with the `semantic` query

The `semantic` query can be used as a part of a hybrid search where the `semantic` query is combined with lexical queries.

> For example, the query below finds documents with title matching "mountain lake", and combines them with results from a semantic search on the field `title_semantic`, that is a `semantic_text` field.

Reviewer 2: ```suggestion
For example, the query below finds documents with the `title` field matching "mountain lake", and combines them with results from a semantic search on the field `title_semantic`, that is a `semantic_text` field.
```

The combined documents are then scored, and the top 3 top scored documents are returned.

[source,console]
------------------------------------------------------------
POST my-index/_search
{
  "size" : 3,
  "query": {
    "bool": {
      "should": [
        {
          "match": {
            "title": {
              "query": "mountain lake",
              "boost": 1
            }
          }
        },
        {
          "semantic": {
            "field": "title_semantic",
            "query": "mountain lake",
            "boost": 2
          }
        }
      ]
    }
  }
}
------------------------------------------------------------
// TEST[skip:TBD]

> You can also use semantic_text as part of <<rrf,Reciprocal Rank Fusion>>:

Reviewer 2: ```suggestion
You can also use semantic_text as part of <<rrf,Reciprocal Rank Fusion>> to make ranking relevant results easier:
```

[source,console]
------------------------------------------------------------
GET my-index/_search
{
  "retriever": {
    "rrf": {
      "retrievers": [
        {
          "standard": {
            "query": {
              "term": {
                "text": "shoes"
              }
            }
          }
        },
        {
          "semantic": {
            "field": "semantic_field",
            "query": "shoes"
          }
        }
      ],
      "rank_window_size": 50,
      "rank_constant": 20
    }
  }
}
------------------------------------------------------------
// TEST[skip:TBD]

[discrete]
[[advanced-search]]
=== Advanced search on `semantic_text` fields

The `semantic` query uses default settings for searching on `semantic_text` fields for ease of use.
If you want to fine-tune a search on a `semantic_text` field, you need to know the task type used by the `inference_id` configured in `semantic_text`.
You can find the task type using the <<get-inference-api>>, and check the `task_type` associated with the {infer} service.

> Depending on the `task_type`, use either the `sparse_vector` or the `knn` query for greater flexibility and customization.

Reviewer 2: ```suggestion
Depending on the `task_type`, use either the <<query-dsl-sparse-vector-query,`sparse_vector`> or the <<query-dsl-knn-query,`knn`>> query for greater flexibility and customization.
```

------------------------------------------------------------
// TEST[skip:TBD]

You can customize the `sparse_vector` query to include specific settings, like <<sparse-vector-query-with-pruning-config-and-rescore-example,pruning configuration>>.

[discrete]
[[search-text-inferece]]
==== Search with `text_embedding` inference

> When the {infer} endpoint uses a `text_embedding` model, you can use a <<query-dsl-knn-query,`knn`query>> on a `semantic_text` field in the following way:

Reviewer: The "`knn` query" link is not rendered properly here
